File: app/src/main/java/com/drdisagree/iconify/config/XPrefs.java
package com.drdisagree.iconify.config;

/* Modified from AOSPMods
 * https://github.com/siavash79/AOSPMods/blob/canary/app/src/main/java/sh/siava/AOSPMods/XPrefs.java
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

import static com.drdisagree.iconify.common.Resources.SharedXPref;
import static de.robv.android.xposed.XposedBridge.log;

import android.content.Context;
import android.content.SharedPreferences;

import com.crossbowffs.remotepreferences.RemotePreferences;
import com.drdisagree.iconify.BuildConfig;
import com.drdisagree.iconify.xposed.HookEntry;
import com.drdisagree.iconify.xposed.ModPack;

public class XPrefs {

    private static final SharedPreferences.OnSharedPreferenceChangeListener listener = (sharedPreferences, key) -> loadEverything(key);
    public static SharedPreferences Xprefs;
    private static String packageName;

    public static void init(Context context) {
        packageName = context.getPackageName();
        Xprefs = new RemotePreferences(context, BuildConfig.APPLICATION_ID, SharedXPref, true);
        log("Iconify Version: " + BuildConfig.VERSION_NAME);
        Xprefs.registerOnSharedPreferenceChangeListener(listener);
    }

    public static void loadEverything(String... key) {
        for (ModPack thisMod : HookEntry.runningMods) {
            thisMod.updatePrefs(key);
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/config/XPrefs.kt
package com.drdisagree.iconify.config

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import com.crossbowffs.remotepreferences.RemotePreferences
import com.drdisagree.iconify.BuildConfig
import com.drdisagree.iconify.common.Const.PREF_UPDATE_EXCLUSIONS
import com.drdisagree.iconify.common.Resources.SHARED_XPREFERENCES
import com.drdisagree.iconify.xposed.HookEntry

object XPrefs {

    @JvmField
    var Xprefs: SharedPreferences? = null
    private var packageName: String? = null
    private val listener =
        OnSharedPreferenceChangeListener { _: SharedPreferences?, key: String? ->
            loadEverything(
                key
            )
        }

    @JvmStatic
    fun init(context: Context) {
        packageName = context.packageName
        Xprefs = RemotePreferences(context, BuildConfig.APPLICATION_ID, SHARED_XPREFERENCES, true)
        (Xprefs as RemotePreferences).registerOnSharedPreferenceChangeListener(listener)
    }

    private fun loadEverything(vararg key: String?) {
        if (key.isEmpty() || key[0].isNullOrEmpty() || PREF_UPDATE_EXCLUSIONS.any { exclusion ->
                key[0]?.equals(exclusion) == true
            }) {
            return
        }

        HookEntry.runningMods.forEach { thisMod ->
            thisMod.updatePrefs(*key.filterNotNull().toTypedArray())
        }
    }
}
